shared/team-building/service-tab-bar.desktop.tsx

```diff
@@ -4,76 +4,156 @@ import * as Styles from '../styles'
+    }>
+  ) => (
+    <>
+      <Kb.Box2
+        direction="vertical"
+        style={styles.moreNetworks1}
+        fullHeight={true}
+        centerChildren={true}
+        ref={props.setAttachmentRef}
+      >
+        <Kb.WithTooltip text="More networks" containerStyle={styles.moreNetworks2}>
+          <Kb.ClickableBox onClick={props.toggleShowingMenu} style={styles.moreNetworks3}>
+            <Kb.Text type="BodyBigExtrabold" style={styles.moreText}>
+              •••
+            </Kb.Text>
+          </Kb.ClickableBox>
+        </Kb.WithTooltip>
+      </Kb.Box2>
+      <Kb.FloatingMenu
+        attachTo={props.getAttachmentRef}
+        closeOnSelect={true}
+        items={props.services.map(service => ({
+          onClick: () => props.onChangeService(service),
+          title: service,
+          view: <MoreNetworkItem service={service} />,
+        }))}
+        onHidden={props.toggleShowingMenu}
+        visible={props.showingMenu}
+      />
+    </>
+  )
+)
+
+const MoreNetworkItem = (props: {service: ServiceIdWithContact}) => (
+  <Kb.Box2 direction="horizontal" fullHeight={true} alignItems="center">
+    <Kb.Icon
+      style={{marginRight: Styles.globalMargins.tiny}}
+      fontSize={16}
+      type={serviceIdToIconFont(props.service)}
     />
-  </Kb.ClickableBox>
+    <Kb.Text type="Body">{serviceIdToLongLabel(props.service)}</Kb.Text>
+  </Kb.Box2>
 )
 
 const undefToNull = (n: number | undefined | null): number | null => (n === undefined ? null : n)
 
-export const ServiceTabBar = (props: Props) => (
-  <Kb.Box2 direction="horizontal" fullWidth={true} style={styles.tabBarContainer}>
-    {Constants.services.map(service => (
-      <ServiceIcon
-        key={service}
-        service={service}
-        label={serviceIdToLongLabel(service)}
-        labelPresence={1}
-        onClick={() => props.onChangeService(service)}
-        count={undefToNull(props.serviceResultCount[service])}
-        showCount={props.showServiceResultCount}
-        isActive={props.selectedService === service}
-      />
-    ))}
-  </Kb.Box2>
-)
+export const ServiceTabBar = (props: Props) => {
+  const [lastSelectedUnlockedService, setState] = React.useState<ServiceIdWithContact | null>(null)
+  const nLocked = 3 // Services always out front on the left. Add one to get the number out front.
+  const lockedServices = Constants.services.slice(0, nLocked)
+  let frontServices = lockedServices
+  if (Constants.services.indexOf(props.selectedService) < nLocked) {
+    // Selected service is locked
+    if (lastSelectedUnlockedService === null) {
+      frontServices = lockedServices.concat(Constants.services.slice(nLocked, nLocked + 1))
+    } else {
+      frontServices = lockedServices.concat([lastSelectedUnlockedService])
+    }
+  } else {
+    frontServices = lockedServices.concat([props.selectedService])
+    if (lastSelectedUnlockedService !== props.selectedService) {
+      setState(props.selectedService)
```

Comment:
We don't really ever want to setState or do other side effects in the render part of a component. What would you think handling this in the `onClick` callback of `ServiceIcon`?

Example:
```tsx
const onServiceIconClick = React.useCallback(service => {
  if (!lockedServices.includes(service)) {
    setSelectedUnlockedService(service)
  }
  props.onChangeService(service)
}, ...)
```

I realize the typing for this probably isn't correct, but hopefully that's enough of a starting point. Feel free to ping me too!

Reply:
What's the issue with setState in render? Loops?

That looks nice. One thing about it, that wouldn't actually come up in the current setup, is that it would miss changes to props.selectedService that happened due to some other trigger.

What do you think about something like this?
```
React.useEffect(
  () = setSelectedUnlockedService(props.selectedService),
  [props.selectedService])
```

Reply:
From the [React docs](https://reactjs.org/docs/react-component.html#render):

> The render() function should be pure, meaning that it does not modify component state, it returns the same result each time it’s invoked, and it does not directly interact with the browser.

The main worry is infinite loops, since `setState` triggers a re-render meaning you're borked if the same code path evaluates twice.